src/cargo/core/compiler/context/mod.rs

```diff
@@ -258,6 +258,16 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
 
                 self.compilation.to_doc_test.push(compilation::Doctest {
                     unit: unit.clone(),
+                    artifact_meta: self
+                        .unit_deps(unit)
+                        .iter()
+                        .filter_map(|dep| {
+                            dep.unit
+                                .artifact
+                                .is_true()
+                                .then(|| self.files().metadata(&dep.unit))
+                        })
+                        .collect(),
```

Comment:
Absolutely fantastic, thanks so much! Initially I was skeptical that this could work assuming there must have been _some_ reason for this complexity. However, with every step on the way it became clear that the complexity was in fact accidental, and a simpler implementation is indeed possible :).

The implementation is in this comment: 3a34c2d7af1b70f79687954d3b7637d2f6d0dd7b .